src/Microsoft.ML.CpuMath/AvxIntrinsics.cs

```diff
@@ -185,15 +185,30 @@ public static unsafe void MatMulX(float[] mat, float[] src, float[] dst, int cro
                             Vector256<float> vector = Avx.LoadVector256(pSrcCurrent);
 
                             float* pMatTemp = pMatCurrent;
-                            Vector256<float> x01 = Avx.Multiply(vector, Avx.LoadVector256(pMatTemp));
-                            Vector256<float> x11 = Avx.Multiply(vector, Avx.LoadVector256(pMatTemp += ccol));
-                            Vector256<float> x21 = Avx.Multiply(vector, Avx.LoadVector256(pMatTemp += ccol));
-                            Vector256<float> x31 = Avx.Multiply(vector, Avx.LoadVector256(pMatTemp += ccol));
+                            if (Fma.IsSupported)
+                            {
+                                Vector256<float> x01 = Avx.LoadVector256(pMatTemp);
+                                Vector256<float> x11 = Avx.LoadVector256(pMatTemp += ccol);
+                                Vector256<float> x21 = Avx.LoadVector256(pMatTemp += ccol);
+                                Vector256<float> x31 = Avx.LoadVector256(pMatTemp += ccol);
+
+                                res0 = Fma.MultiplyAdd(vector, x01, res0);
```

Comment:
it would be nice to validate the codegen here and assert that the various loads are being folded into the FMA operation.

Reply:
If this was "normal code", I would normally say we should pull all the calls to `Avx.LoadVector256` above the `if-else` statements, so they aren't duplicated between the two. We are doing that in other methods below.

If we do that, does the codegen still look "acceptable"?

---
In reply to: [226137860](https://github.com/dotnet/machinelearning/pull/1292#discussion_r226137860) [](ancestors = 226137860)